File: flutter/test/chat_test.dart
import 'package:buble/buble.dart';
import 'package:test/test.dart';

import 'test_http_client.dart';

void main() {
  test('creates OpenAI-compatible chat completion', () async {
    final http = TestHttpClient([
      jsonResponse({
        'choices': [
          {
            'message': {'content': 'Hello'},
          },
        ],
      }),
    ]);
    final client = BubleClient(apiKey: 'sk_test', httpClient: http);

    final response = await client.chat.completions.create({
      'model': 'openai/gpt-5.4',
      'messages': [
        {'role': 'user', 'content': 'Hi'},
      ],
    });

    expect(response['choices'], isA<List<Object?>>());
    expect((http.requests.single.jsonBody as Map)['stream'], false);
  });

  test('streams OpenAI text', () async {
    final http = TestHttpClient([
      textStreamResponse([
        'data: {"choices":[{"delta":{"content":"Hel"}}]}\n\n',
        'data: {"choices":[{"delta":{"content":"lo"}}]}\n\n',
        'data: [DONE]\n\n',
      ]),
    ]);
    final client = BubleClient(apiKey: 'sk_test', httpClient: http);

    final stream = await client.chat.completions.streamText({
      'model': 'openai/gpt-5.4',
      'messages': [
        {'role': 'user', 'content': 'Hi'},
      ],
    });

    expect(await stream.join(), 'Hello');
  });

  test('calls Gemini model path', () async {
    final http = TestHttpClient([
      jsonResponse({'candidates': <Object?>[]}),
    ]);
    final client = BubleClient(apiKey: 'sk_test', httpClient: http);

    await client.chat.gemini.generateContent('openai/gpt-5.4', {
      'contents': <Object?>[],
    });

    expect(
      http.requests.single.url.path,
      '/api/v1beta/models/openai/gpt-5.4:generateContent',
    );
  });
}

File: elixir/test/buble/chat_test.exs
defmodule Buble.ChatTest do
  use ExUnit.Case, async: true

  setup do
    {:ok, client} = Buble.Client.new(api_key: "sk-test", transport: Buble.TestTransport)
    %{client: client}
  end

  test "creates OpenAI-compatible chat completions", %{client: client} do
    Process.put(:buble_test_handler, fn _client, method, path, opts ->
      assert method == :post
      assert path == "/api/v1/chat/completions"
      assert Keyword.fetch!(opts, :json)["stream"] == false
      {:ok, %{"choices" => [%{"message" => %{"content" => "Hello"}}]}}
    end)

    assert {:ok, %{"choices" => [_]}} =
             Buble.Chat.Completions.create(client, %{
               model: "chatgpt-5-4",
               messages: [%{role: "user", content: "Hello"}]
             })
  end

  test "streams OpenAI-compatible text", %{client: client} do
    Process.put(:buble_test_handler, fn _client, _method, _path, _opts ->
      {:ok,
       """
       data: {"choices":[{"delta":{"content":"Hel"}}]}

       data: {"choices":[{"delta":{"content":"lo"}}]}

       data: [DONE]

       """}
    end)

    assert {:ok, stream} =
             Buble.Chat.Completions.stream_text(client, %{model: "chatgpt-5-4", messages: []})

    assert Enum.to_list(stream) == ["Hel", "lo"]
  end

  test "calls Anthropic Messages-compatible endpoint", %{client: client} do
    Process.put(:buble_test_handler, fn _client, method, path, opts ->
      assert method == :post
      assert path == "/api/v1/messages"
      assert Keyword.fetch!(opts, :json)["stream"] == false
      {:ok, %{"content" => [%{"text" => "Hello"}]}}
    end)

    assert {:ok, %{"content" => [_]}} =
             Buble.Chat.Messages.create(client, model: "claude", messages: [])
  end

  test "calls Gemini-compatible endpoint", %{client: client} do
    Process.put(:buble_test_handler, fn _client, method, path, opts ->
      assert method == :post
      assert path == "/api/v1beta/models/gemini-3-pro:generateContent"
      assert Keyword.fetch!(opts, :json) == %{"contents" => []}
      {:ok, %{"candidates" => []}}
    end)

    assert {:ok, %{"candidates" => []}} =
             Buble.Chat.Gemini.generate_content(client, "gemini-3-pro", contents: [])
  end
end
